lib/dataSchemaHelpers.ts

```diff
@@ -0,0 +1,223 @@
+
+const schemaDataCache: {[uri: string]: SchemaJson} = {};
+
+export function getDataSchemaDependencies(
+    schema: DataSchemaData,
+    schemaDataMap: {[id: string]: DataSchemaData} = {}
+): DataSchemaData[] {
+    return _.compact(schema.requiredDependencies.map(id => schemaDataMap[id]));
+}
+
+export function getDataSchemaParents(
+    schema: DataSchemaData,
+    schemaDataMap: {[id: string]: DataSchemaData} = {}
+): DataSchemaData[] {
+    return _.compact(schema.parentIds.map(id => schemaDataMap[id]));
+}
+
+export function getDataSchemaValidValues(
+    schema: DataSchemaData,
+    schemaDataMap: {[id: string]: DataSchemaData} = {}
+): DataSchemaData[] {
+    return _.compact(schema.validValues.map(id => schemaDataMap[id]));
+}
+
+export function hasNonEmptyValidValues(schemaData: DataSchemaData[]): boolean {
+    return !_.isEmpty(_.flatten(schemaData.map(s => s.validValues)));
+}
+
+export async function getDataSchema(
+    ids: string[],
+    dataUri: string = DEFAULT_SCHEMA_URL
+): Promise<{
+    dataSchemaData: DataSchemaData[],
+    schemaDataMap: {[id: string]: DataSchemaData}
+}> {
+    const schemaData = getDataSchemaData(await getSchemaData(dataUri));
+    const schemaDataMap = _.keyBy(schemaData, d => d.id);
+    const dataSchemaData = _.compact(ids.map(id => schemaDataMap[id]));
+
+    return {dataSchemaData, schemaDataMap};
+}
+
+export async function getSchemaData(dataUri?: string): Promise<SchemaJson> {
+    if (!dataUri) {
+        // return {
+        //     '@context': defaultSchema['@context'],
+        //     '@graph': defaultSchema['@graph'],
+        //     '@id': defaultSchema['@id'],
+        // } as SchemaJson;
+
+        return DEFAULT_SCHEMA;
+    }
+
+    // do not fetch again if fetched before
+    if (schemaDataCache[dataUri]) {
+        return schemaDataCache[dataUri];
+    }
+
+    try {
+        // https://raw.githubusercontent.com/Sage-Bionetworks/schematic/main/data/schema_org_schemas/example.jsonld
+        // https://github.com/ncihtan/hsim/blob/master/schema/HTAN.jsonld
+        const res = await fetch(dataUri);
+
+        // const json = await res.json();
+        const text = await res.text();
+        const json = JSON.parse(text);
+
+        // cache the schema data
+        schemaDataCache[dataUri] = json;
+
+        return json;
+    } catch {
+        // console.error(`Invalid Url: ${dataUri}`);
+        return DEFAULT_SCHEMA;
+    }
+}
+
+export function getDataSchemaData(schemaJson?: SchemaJson): DataSchemaData[] {
+    const schema: SchemaData[] = schemaJson ? schemaJson['@graph'] || [] : []
+    const context: SchemaContext = schemaJson
+        ? schemaJson['@context'] || {}
+        : {};
+
+    if (schema.length > 0) {
+        return schema.map(mapSchemaDataToDataSchemaData(context));
+    }
+
+    return [] as DataSchemaData[];
+}
+
+export function mapSchemaDataToDataSchemaData(context: SchemaContext): (nd: SchemaData) => DataSchemaData {
```

Comment:
lol this function name... can it be improved? maybe not

Reply:
let's keep the original name for now :) https://github.com/onursumer/Synapse-React-Client/blob/01fb7e594517f9f4b4e9950fd0bd21e45e17eeb9/src/lib/containers/data_schema/utils/mapSchemaDataToDataSchemaData.ts#L8

Ideally we should be able to import these functions, but there is no npm library that we can use atm.